lib/services/authority.go

```diff
@@ -50,6 +69,105 @@ func NewCertAuthority(
 	})
 }
 
+// ValidateCertAuthority validates the CertAuthority
+func ValidateCertAuthority(ca CertAuthority) (err error) {
+	if err = ca.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	switch ca.GetType() {
+	case UserCA, HostCA:
+		err = checkUserOrHostCA(ca)
+	case types.JWTSigner:
+		err = checkJWTKeys(ca)
+	}
+	return trace.Wrap(err)
+}
+
+func checkUserOrHostCA(ca CertAuthority) error {
+	if len(ca.GetCheckingKeys()) == 0 {
+		return trace.BadParameter("certificate authority missing SSH public keys")
+	}
+	if len(ca.GetTLSKeyPairs()) == 0 {
+		return trace.BadParameter("certificate authority missing TLS key pairs")
+	}
+	if _, err := ca.Checkers(); err != nil {
+		return trace.Wrap(err)
+	}
+	if _, err := ca.Signers(); err != nil {
+		return trace.Wrap(err)
+	}
+	// This is to force users to migrate
+	if len(ca.GetRoles()) != 0 && len(ca.GetRoleMap()) != 0 {
+		return trace.BadParameter("should set either 'roles' or 'role_map', not both")
+	}
+	err := ca.GetRoleMap().Check()
+	return trace.Wrap(err)
+}
+
+func checkJWTKeys(ca CertAuthority) error {
+	// Check that some JWT keys have been set on the CA.
+	if len(ca.GetJWTKeyPairs()) == 0 {
+		return trace.BadParameter("missing JWT CA")
+	}
+
+	var err error
+	var privateKey crypto.Signer
+
+	// Check that the JWT keys set are valid.
+	for _, pair := range ca.GetJWTKeyPairs() {
+		if len(pair.PrivateKey) > 0 {
+			privateKey, err = utils.ParsePrivateKey(pair.PrivateKey)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+		publicKey, err := utils.ParsePublicKey(pair.PublicKey)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		cfg := &jwt.Config{
+			Algorithm:   defaults.ApplicationTokenAlgorithm,
+			ClusterName: ca.GetClusterName(),
+			PrivateKey:  privateKey,
+			PublicKey:   publicKey,
+		}
+		if _, err = jwt.New(cfg); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+// GetJWTSigner returns the active JWT key used to sign tokens.
+func GetJWTSigner(ca CertAuthority, config jwt.Config) (*jwt.Key, error) {
```

Comment:
Why not pass the clock in here to be consistent with how you did this elsewhere?